openedx/features/course_experience/views/course_outline.py

```diff
@@ -30,21 +32,75 @@ def render_to_fragment(self, request, course_id=None, page_context=None, **kwarg
         if not course_block_tree:
             return None
 
-        content_milestones = self.get_content_milestones(request, course_key)
-
         context = {
             'csrf': csrf(request)['csrf_token'],
             'course': course_overview,
-            'blocks': course_block_tree,
-            'gated_content': content_milestones
+            'blocks': course_block_tree
         }
-        html = render_to_string('course_experience/course-outline-fragment.html', context)
-        return Fragment(html)
+
+        # TODO: EDUCATOR-2283 Remove this check when the waffle flag is turned on in production
+        if waffle.new_course_outline_enabled(course_key=course_key):
+            xblock_display_names = self.create_xblock_id_and_name_dict(course_block_tree)
+
+            gated_content = self.get_content_milestones(request, course_key)
+
+            context['gated_content'] = gated_content
+            context['xblock_display_names'] = xblock_display_names
+
+            # TODO: EDUCATOR-2283 Rename this file to course-outline-fragment.html
+            html = render_to_string('course_experience/course-outline-fragment-new.html', context)
+            return Fragment(html)
+        else:
+            content_milestones = self.get_content_milestones_old(request, course_key)
+
+            context['gated_content'] = content_milestones
+
+            # TODO: EDUCATOR-2283 Remove this file
+            html = render_to_string('course_experience/course-outline-fragment-old.html', context)
+            return Fragment(html)
+
+    def create_xblock_id_and_name_dict(self, course_block_tree, xblock_display_names=None):
+        if xblock_display_names is None:
+            xblock_display_names = {}
+
+        if course_block_tree.get('id'):
+            xblock_display_names[course_block_tree['id']] = course_block_tree['display_name']
+
+        if course_block_tree.get('children'):
+            for child in course_block_tree['children']:
+                self.create_xblock_id_and_name_dict(child, xblock_display_names)
+
+        return xblock_display_names
 
     def get_content_milestones(self, request, course_key):
         """
         Returns dict of subsections with prerequisites and whether the prerequisite has been completed or not
         """
+        def _get_key_of_prerequisite(namespace):
+            return re.sub('.gating', '', namespace)
+
+        all_course_milestones = get_course_content_milestones(course_key)
+
+        uncompleted_prereqs = [
```

Comment:
You should probably make this a set (by changing the `[` to `{`) if you're going to use it to look things up.